linkerd/app/inbound/src/http/router.rs

```diff
@@ -417,9 +418,60 @@ impl tap::Inspect for Logical {
     }
 
     fn route_labels<B>(&self, req: &http::Request<B>) -> Option<tap::Labels> {
-        req.extensions()
+        let mut labels = std::collections::BTreeMap::new();
+        for (k, v) in req
+            .extensions()
             .get::<profiles::http::Route>()
-            .map(|r| r.labels().clone())
+            .iter()
+            .flat_map(|route| route.labels().iter())
+        {
+            labels.insert(k.clone(), v.clone());
+        }
+        if let Some(route_labels) = req.extensions().get::<RouteLabels>() {
+            labels.insert(
+                "srv_group".to_string(),
+                route_labels.server.0.group().to_string(),
+            );
+            labels.insert(
+                "srv_kind".to_string(),
+                route_labels.server.0.kind().to_string(),
+            );
+            labels.insert(
+                "srv_name".to_string(),
+                route_labels.server.0.name().to_string(),
+            );
+            labels.insert(
+                "route_group".to_string(),
+                route_labels.route.group().to_string(),
+            );
+            labels.insert(
+                "route_kind".to_string(),
+                route_labels.route.kind().to_string(),
+            );
+            labels.insert(
+                "route_name".to_string(),
+                route_labels.route.name().to_string(),
+            );
+        }
+        if let Some(authz_labels) = req.extensions().get::<RouteAuthzLabels>() {
+            labels.insert(
+                "authz_group".to_string(),
+                authz_labels.authz.group().to_string(),
+            );
+            labels.insert(
+                "authz_kind".to_string(),
+                authz_labels.authz.kind().to_string(),
+            );
+            labels.insert(
+                "authz_name".to_string(),
+                authz_labels.authz.name().to_string(),
+            );
+        }
+        if labels.is_empty() {
+            None
+        } else {
+            Some(std::sync::Arc::new(labels))
+        }
```

Comment:
We already do this work here:

https://github.com/linkerd/linkerd2-proxy/blob/92d50fcfc5b9b7e0c71049d9250ed1e43fab1da4/linkerd/app/inbound/src/http/router.rs#L250-L276

That is, all of the copying/allocation happens once per stack and not for each request. This also avoids messing with extensions, etc. IIUC, we're already including route labels there...